### ISSUE TITLE
re.Match.group() output doesn't raise type error

### ISSUE THREAD

**Issue by john-dicarlo** [python/mypy#16441]
```python
# example.py

import re

def foo(s: str) -> str:
    print(s)
    return s

m: re.Match[str] | None = re.match('(a)(b)?', 'a')

if m is not None:
    foo(None) # Argument 1 to "foo" has incompatible type "None"; expected "str"  [arg-type] # as expected
    
    reveal_type(m.group(2))  # Revealed type is "Union[builtins.str, Any]"
    foo(m.group(2)) # does not raise an [arg-type] error
    assert m.group(2) is None # passes

```

```bash
$ mypy --strict example.py 
example.py:12: error: Argument 1 to "foo" has incompatible type "None"; expected "str"  [arg-type]
example.py:14: note: Revealed type is "Union[builtins.str, Any]"
Found 1 error in 1 file (checked 1 source file)
```

Why doesn't `m.group(2)` raise an [arg-type] error here? 

__Your Environment__

mypy 1.6.1 (compiled: yes)
Python 3.10.12

**Comment by hauntsaninja** [comment]
See https://github.com/python/typeshed/issues/9465#issuecomment-1373844980 / https://github.com/python/typeshed/pull/9482#issuecomment-1375989845

There's a balance between false negatives (like this) and false positives. Typeshed aims to avoid false positives, which is why the return type uses a union containing Any. For many — maybe even most — regular expressions, group can't actually be None, so this is a good decision.

**Comment by KotlinIsland** [comment]
Basedmypy works correctly in these cases:
```
import re

s: str
if m := re.match("a(b)?(c)", s):
    reveal_type(m.groups())  #  (str | None, str)
```
